File: fp-in-ocaml/answers/errorhandling/validated.ml
(* [either.ml] に書いた通り、OCaml での right-biased な型は {{:https://ocaml.org/manual/5.4/api/Result.html}[Stdlib.Result]} になる。
   ここでは、オリジナルに沿って、[Validated] を定義していく。 *)

type (+'e, +'a) t = Valid of 'a | Invalid of 'e list

let to_either = function
  | Valid a -> Either.Right a
  | Invalid es -> Either.Left es

let map f = function Valid a -> Valid (f a) | invalid -> invalid

(* map2 のための helper *)
let get_invalids = function Invalid es -> es | Valid _ -> []

let map2 f = function
  | Valid a, Valid b -> Valid (f a b)
  | a, b -> Invalid (get_invalids a @ get_invalids b)

(* Scala 版では [fromEither] だが、OCaml の文化では [of_] が多いので修正 *)
let of_either = function
  | Either.Right a -> Valid a
  | Either.Left es -> Invalid es

let traverse f =
  (* 通常のモジュールだけでなく、ラベル付き引数を使える [~Labels] モジュールが存在する。
     引数の順序で悩まなくていいので、特に [fold_] 系の関数で便利。
   *)
  ListLabels.fold_right
    ~f:(fun a acc -> map2 List.cons (f a, acc))
    ~init:(Valid [])

(** @see <https://ocaml.org/manual/5.4/polymorphism.html#ss:valuerestriction>
      6.1.2 The value restriction *)
let sequence x = traverse Fun.id x

module More_general = struct
  type (+'e, +'a) t = Valid of 'a | Invalid of 'e

  let to_either = function
    | Valid a -> Either.Right a
    | Invalid es -> Either.Left es

  let map f = function Valid a -> Valid (f a) | invalid -> invalid

  (* map2 のための helper *)
  let get_invalids = function Invalid e -> [ e ] | Valid _ -> []

  let map2 ~combine_errors f = function
    | Valid a, Valid b -> Valid (f a b)
    | a, b -> Invalid (combine_errors (get_invalids a) (get_invalids b))

  (* Scala 版では [fromEither] だが、OCaml の文化では [of_] が多いので修正 *)
  let of_either = function
    | Either.Right a -> Valid a
    | Either.Left es -> Invalid es

  let traverse ~combine_errors f =
    ListLabels.fold_right
      ~f:(fun a acc -> map2 ~combine_errors List.cons (f a, acc))
      ~init:(Valid [])

  let sequence = traverse Fun.id
end

File: fp-in-haskell/src/FpInHaskell/Answers/ErrorHandling/Validated.hs
module FpInHaskell.Answers.ErrorHandling.Validated (
    Validated (..),
    toEither,
    fromEither,
    map,
    map2,
    traverse,
    sequence,
) where

import FpInHaskell.Answers.ErrorHandling.Either (Either (Left, Right))
import Prelude hiding (Either (..), map, sequence, traverse)

-- `Validated` は `Either` に準じた型だが、`Invalid` が誤りを1つではなく**リストで**保持する点が異なる。
-- `map2` が両辺とも失敗のときにリストを連結することで、`Either.map2` と違って複数の誤りを
-- 蓄積できる(`Either.map2` は最初に出会った1つの誤りで打ち切ってしまう)。
-- この型自体には演習がなく(原典でも全関数が提供済み)、`Either` を補う参考実装として
-- Answers 側にのみ用意する。
data Validated e a
    = Valid a
    | Invalid [e]
    deriving (Show, Eq)

toEither :: Validated e a -> Either [e] a
toEither (Valid x) = Right x
toEither (Invalid es) = Left es

fromEither :: Either [e] a -> Validated e a
fromEither (Right x) = Valid x
fromEither (Left es) = Invalid es

map :: (a -> b) -> Validated e a -> Validated e b
map _ (Invalid es) = Invalid es
map f (Valid x) = Valid (f x)

-- 両辺が `Valid` のときだけ `f` を適用する。両辺とも `Invalid` なら、誤りのリストを連結して蓄積する
-- (`Either.map2` のように最初の誤りだけを返すのではない点に注意)。
map2 :: (a -> b -> c) -> Validated e a -> Validated e b -> Validated e c
map2 f (Valid x) (Valid y) = Valid (f x y)
map2 _ (Invalid es) (Valid _) = Invalid es
map2 _ (Valid _) (Invalid es) = Invalid es
map2 _ (Invalid es1) (Invalid es2) = Invalid (es1 ++ es2)

traverse :: (a -> Validated e b) -> [a] -> Validated e [b]
traverse _ [] = Valid []
traverse f (h : t) = map2 (:) (f h) (traverse f t)

sequence :: [Validated e a] -> Validated e [a]
sequence = traverse id
